cloud/aws/services/ec2/natgateway.go

```diff
@@ -0,0 +1,126 @@
+package ec2
+
+import (
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/pkg/errors"
+	"sigs.k8s.io/cluster-api-provider-aws/cloud/aws/providerconfig/v1alpha1"
+)
+
+func (s *Service) reconcileNatGateways(subnets []*v1alpha1.Subnet, vpc *v1alpha1.VPC) error {
+	if !s.containsPrivateSubnets(subnets) {
+		return nil
+	}
+
+	existing, err := s.describeNatGateways(vpc.ID)
+	if err != nil {
+		return err
+	}
+
+	for _, sn := range subnets {
+		if sn.ID == "" || !sn.IsPublic {
+			continue
+		}
+		if _, ok := existing[sn.ID]; ok {
+			continue
+		}
+
+		_, err := s.createNatGateway(sn.ID)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *Service) describeNatGateways(vpcID string) (map[string]*ec2.NatGateway, error) {
+	describeNatGatewayInput := ec2.DescribeNatGatewaysInput{
+		Filter: []*ec2.Filter{
+			{
+				Name:   aws.String("vpc-id"),
+				Values: []*string{aws.String(vpcID)},
+			},
+		},
+	}
+
+	gateways := make(map[string]*ec2.NatGateway)
+
+	err := s.ec2.DescribeNatGatewaysPages(
+		&describeNatGatewayInput,
+		func(page *ec2.DescribeNatGatewaysOutput, lastPage bool) bool {
+			for _, r := range page.NatGateways {
+				gateways[*r.SubnetId] = r
+			}
+			return false
+		})
+
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to describe NAT gateways with VPC ID %q", vpcID)
+	}
+	return gateways, nil
+}
+
+func (s *Service) allocateAddress() (string, error) {
+	out, err := s.ec2.AllocateAddress(&ec2.AllocateAddressInput{})
+
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to create Elastic IP address")
+	}
+
+	return *out.AllocationId, nil
+}
+
+func (s *Service) createNatGateway(subnetID string) (*ec2.NatGateway, error) {
+
+	ip, err := s.allocateAddress()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to create IP address for NAT gateway for subnet ID %q", subnetID)
+	}
+
+	out, err := s.ec2.CreateNatGateway(&ec2.CreateNatGatewayInput{
+		SubnetId:     aws.String(subnetID),
+		AllocationId: aws.String(ip),
+	})
+
+	final, err := s.waitUntilNatGatewayAvailable(*out.NatGateway.NatGatewayId)
+
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to create NAT gateway for subnet ID %q", subnetID)
+	}
+
+	return final, nil
+}
+
+func (s *Service) waitUntilNatGatewayAvailable(gatewayID string) (*ec2.NatGateway, error) {
+
+	err := s.ec2.WaitUntilNatGatewayAvailable(&ec2.DescribeNatGatewaysInput{
```

Comment:
Do we need a helper method for this?

Reply:
Wasn't sure either, when I wrote it earlier, so it's been inlined.